contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+        // slither-disable-next-line unused-return return value irrelevant, address guaranteed removed
+        schedule.tokenHolders.remove(from);
+      }
+    }
+  }
+
+  /**
+   * For each index-matched account, removal id, and amount, revokes amount of tokens from schedule corresponding
+   * to the specified removal and transfers to the corresponding account.
+   *
+   * @dev Transfers any unreleased tokens in the removal id's corresponding schedule and reduces the total supply
+   * of that token. No change is made to balances that have released but not yet been claimed.
+   * If a token has multiple owners, balances are burned proportionally to ownership percentage, summing to the total
+   * amount being revoked.
+   *
+   * The behavior of this function can be used in two specific ways:
+   * - To revoke a specific number of tokens as specified by the `amounts` array.
+   * - To revoke all remaining revokable tokens in a schedule set amount to 0 in the `amounts` array.
+   *
+   * ##### Requirements:
+   *
+   * - Can only be used when the caller has the `TOKEN_REVOKER_ROLE`
+   * - The requirements of _beforeTokenTransfer apply to this function
+   * - toAccounts.length == removalIds.length == amounts.length
+   */
+  function batchRevokeUnreleasedTokenAmounts(
+    address[] calldata toAccounts,
+    uint256[] calldata removalIds,
+    uint256[] calldata amounts
+  ) external whenNotPaused onlyRole(TOKEN_REVOKER_ROLE) {
+    if (!(toAccounts.length == removalIds.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "removalIds"
+      });
+    }
+    if (!(toAccounts.length == amounts.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "amounts"
+      });
+    }
+    for (uint256 i = 0; i < toAccounts.length; i++) {
+      _revokeUnreleasedTokens(toAccounts[i], removalIds[i], amounts[i]);
+    }
+  }
+
+  // Private implementations ==========================================
+
+  /**
+   * Wraps minting of wrapper token and schedule setup.
+   *
+   * @dev If no schedule is set up for the specified removal id, one is created.
+   *
+   * @param removalId uint256 The removal for which funds are being deposited.
+   * @param amount uint256 Quantity of `_bridgedPolygonNori` to deposit
+   * @param userData bytes extra information provided by the user (if any)
+   * @param operatorData bytes extra information provided by the operator (if any)
+   */
+  function _depositFor(
+    uint256 removalId,
+    uint256 amount,
+    bytes memory userData,
+    bytes memory operatorData
+  ) internal returns (bool) {
+    uint256 projectId = _removal.getProjectIdForRemoval(removalId);
+    Removal.ScheduleData memory scheduleData = _removal
+      .getScheduleDataForProjectId(projectId);
+    address recipient = scheduleData.supplierAddress;
+    if (!_scheduleIdToScheduleStruct[projectId].exists) {
+      _createSchedule(projectId);
+    }
```

Comment:
is this possible?

Reply:
yeah so it's very possible that removals will be listed for sale (which is the point at which this function gets called) that belong to a schedule for which removals had already been listed and therefore the schedule had already been created.  so we just want to make sure that before we call `_createSchedule` for this batch being listed that we don't already have the schedule set up or else the transaction will revert

Reply:
Sorry, asking something slightly different. 

Is it ever possible that a schedule _**doesnt**_ exist for a given removal? I thought we planned on creating schedules for removals when we list them in the market? This call to create a schedule however is happening in the flow of `_depositFor` which is invoked by `tokensReceived` which is invoked when a buyer purchases removals from the market. The fact that this exists here implies it's possible for a removal to be purchased before a removal ever had a schedule, doesn't it?

Reply:
ah yeah, exactly.  It really shouldn't be possible.  this check is there partly as a holdover from `LockedNORI`, and I left it because even in the exceptional case that a removal doesn't have a schedule, since we have enough information in the removal id itself to set up a schedule, it seemed like we should do that because this would be a purchase transaction and we really want to avoid at all costs reverting and wasting a buyer's gas.  However I'd be comfortable removing it entirely, or changing the body of the conditional to just revert if there's no schedule, since this really should never be possible.  what do you think?

Reply:
My vote is to remove it-- seems like if the schedule didnt exist, something really went wrong in the market lifecycle and so falling back to this might be a dangerous default as we should probably be pausing the contracts and looking into what happened

Reply:
remove or revert works for me. It might already revert further down without needing to add overhead of another check, so check that first